crates/sui-framework/sources/dynamic_field.move

```diff
@@ -0,0 +1,136 @@
+}
+
+/// Mutably borrows the `object`s dynamic field with the name specified by `name: Name`.
+/// Aborts with `EFieldDoesNotExist` if the object does not have a field with that name.
+/// Aborts with `EFieldTypeMismatch` if the field exists, but the value does not have the specified
+/// type.
+public fun borrow_mut<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): &mut Value {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    assert!(option::is_some(&field.value), EFieldDoesNotExist);
+    option::borrow_mut(&mut field.value)
+}
+
+/// Removes the `object`s dynamic field with the name specified by `name: Name` and returns the
+/// bound value.
+/// Aborts with `EFieldDoesNotExist` if the object does not have a field with that name.
+/// Aborts with `EFieldTypeMismatch` if the field exists, but the value does not have the specified
+/// type.
+public fun remove<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): Value {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    assert!(option::is_some(&field.value), EFieldDoesNotExist);
+    option::extract(&mut field.value)
+}
+
+// TODO implement exists (without the Value type) once we have lamport timestamps
+/// Returns true if and only if the `object` has a dynamic field with the name specified by
+/// `name: Name`.
+public fun exists_with_type<Name: copy + drop + store, Value: store>(
+    object: &UID,
+    name: Name,
+): bool {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    if (!has_child_object_with_ty<Field<Name, Value>>(object_addr, hash)) return false;
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    option::is_some(&field.value)
+}
+
+public(friend) native fun hash_type_and_key<K: copy + drop + store>(parent: address, k: K): address;
+/// throws `EFieldAlreadyExists` if a child already exists with that ID
+public(friend) native fun add_child_object<Child: key>(parent: address, child: Child);
+/// throws `EFieldDoesNotExist` if a child does not exist with that ID
+/// or throws `EFieldTypeMismatch` if the type does not match
+public(friend) native fun borrow_child_object<Child: key>(parent: address, id: address): &mut Child;
+/// throws `EFieldDoesNotExist` if a child does not exist with that ID
+/// /// or throws `EFieldTypeMismatch` if the type does not match
+public(friend) native fun remove_child_object<Child: key>(parent: address, id: address): Child;
+public(friend) native fun has_child_object(parent: address, id: address): bool;
+public(friend) native fun has_child_object_with_ty<Child: key>(parent: address, id: address): bool;
```

Comment:
we might want to duplicate these in `dynamic_object_field.move` if we want the abort codes to make a little bit more sense, but I think it might be fine as is? unsure

Reply:
I like your single copy + `friend` approach.